File: glimpse/backends/cython_backend/src/array.cpp
/*******************************************************************************
 * Copyright (c) 2011 Mick Thomure                                             *
 * All rights reserved.                                                        *
 *                                                                             *
 * Please see the file COPYING in this distribution for usage terms.           *
 ******************************************************************************/

#include "array.h"
#include <cmath>

Index2D::Index2D(int w0, int w1) : w0(w0), w1(w1) {
}

Index3D::Index3D(int w0, int w1, int w2) : w0(w0), w1(w1), w2(w2) {
}

Index4D::Index4D(int w0, int w1, int w2, int w3) : w0(w0), w1(w1), w2(w2),
    w3(w3) {
}

Index5D::Index5D(int w0, int w1, int w2, int w3, int w4) : w0(w0), w1(w1),
    w2(w2), w3(w3), w4(w4) {
}

// Normalize each location to have constant energy.
void CNormalizeArrayAcrossBand_UnitNorm(ArrayRef3D<float>* data) {
    int num_bands = data->w0;
    // Normalizing an array with a single band will zero the array,
    // since the value at each location is equal to the mean.
    if (num_bands < 2) {
        return;
    }
    int height = data->w1;
    int width = data->w2;
    for (int j = 0; j < height; ++j) {
        for (int i = 0; i < width; ++i) {
            float sum = 0.0;
            for (int b = 0; b < num_bands; ++b) {
                sum += (*data)(b, j, i);
            }
            float mean = sum / num_bands;
            float centered_squares = 0;
            for (int b = 0; b < num_bands; b++) {
                float& v = (*data)(b, j, i);
                v -= mean;
                centered_squares += v * v;
            }
            float norm = sqrt(centered_squares);
            // Only scale if norm is big enough. This avoids amplifying noise.
            if (norm > 1.0) {
                for (int b = 0; b < num_bands; ++b) {
                    (*data)(b, j, i) /= norm;
                }
            }
        }
    }
}

File: glimpse/util/array.py

# Copyright (c) 2011 Mick Thomure
# All rights reserved.
#
# Please see the file COPYING in this distribution for usage terms.

#
# Functions for dealing with N-dimensional arrays.
#

import Image
import math
import numpy

ACTIVATION_DTYPE = numpy.float32

def ArgMax(array):
  """Short-hand to find array indices containing the maximum value."""
  return numpy.transpose(numpy.nonzero(array == array.max()))

def ArgMin(array):
  """Short-hand to find array indices containing the minimum value."""
  return numpy.transpose(numpy.nonzero(array == array.min()))

def ScaleUnitNorm(x):
  """Scale elements of vector (in place), such that result has unit norm."""
  norm = numpy.linalg.norm(x)
#  norm = math.sqrt((x**2).sum())
  if norm == 0:
    x[:] = 1.0 / math.sqrt(x.size)
  else:
    x /= norm
  return x

def ArrayToGreyscaleImage(array, normalize = True):
  """Create a greyscale image from a 2D numpy array. Assumes range of input
     values contains 0."""
  if array.dtype != numpy.float32:
    array = array.astype(numpy.float32)
  if len(array.shape) > 2:
    # Stack bands vertically.
    array = array.reshape((-1, array.shape[-1]))
  if normalize:
    # Normalize array values to lie in [0, 255].
    max_val = max(abs(array.min()), array.max())
    # Map values to lie in [-.5, .5]
    array = array / (2 * max_val)       # makes copy of input
    # Map to [0, 1]
    array += 0.5
    # Map to [0, 255]
    array *= 255
  # Convert to unsigned chars
  array = numpy.asarray(array, dtype = numpy.uint8)
  return Image.fromarray(array, 'L')

def ArrayToRGBImage(array):
  return Image.fromarray(array, 'RGB')

def ArrayListToVector(arrays):
  """Convert list of numpy arrays to a single numpy vector."""
  assert len(arrays) > 0
  out_size = sum(a.size for a in arrays)
  out = numpy.empty((out_size,), arrays[0].dtype)
  offset = 0
  for a in arrays:
    out[offset : offset + a.size] = a.flat
    offset += a.size
  return out

def PadArray(data, out_shape, cval):
  """Pad the border of an array with a constant value."""
  out_shape = numpy.array(out_shape)
  in_shape = numpy.array(data.shape)
  result = numpy.empty(out_shape)
  result[:] = cval
  begin = ((out_shape - in_shape) / 2.0).astype(int)
  result[ [ slice(b, e) for b, e in zip(begin, begin + in_shape) ] ] = data
  return result

def CropArray(data, out_shape):
  """Remove the border of an array.
  data -- input array
  out_shape -- shape of central region to return
  RETURNS view of the central region of input array.
  """
  assert numpy.all(numpy.array(data.shape) >= numpy.array(out_shape))
  out_shape = numpy.array(out_shape)
  in_shape = numpy.array(data.shape)
  begin = ((in_shape - out_shape) / 2.0).astype(int)
  return data[ [ slice(b, e) for b, e in zip(begin, begin + out_shape) ] ]
